### PR TITLE
Adding functional tests for IP Binary DateTime and more String functions

### PR DESCRIPTION
**Pull request by larryluogit**
### Changelog category (leave one):
- Build/Testing/Packaging Improvement

### CODE REVIEW

File: tests/queries/0_stateless/02366_kql_func_string.sql [hunk 1]
```diff
@@ -182,4 +182,28 @@ print '-- indexof (https://docs.microsoft.com/en-us/azure/data-explorer/kusto/qu
...
+print parse_urlquery('k1=v1&k2=v2&k3=v3');
+print '-- strcmp()';
+print strcmp('ABC','ABC'), strcmp('abc','ABC'), strcmp('ABC','abc'), strcmp('abcde','abc');
+print '-- translate()';
+print translate('krasp', 'otsku', 'spark'), translate('abc', '', 'ab'), translate('abc', 'x', 'abc');
+print '-- trim()';
+print trim("--", "--https://www.ibm.com--");
+print '-- trim_start()';
+print trim_start("https://", "https://www.ibm.com");
+print trim_start("[^\w]+", strcat("-  ","Te st", "1", "// $"));
```
Comment (kashwy) [inline tests/queries/0_stateless/02366_kql_func_string.sql:203]: \w  need a escape

File: tests/queries/0_stateless/02366_kql_func_string.sql [hunk 2]
```diff
@@ -182,4 +182,28 @@ print '-- indexof (https://docs.microsoft.com/en-us/azure/data-explorer/kusto/qu
...
+print '-- trim_end()';
+print trim_end("://www.ibm.com", "https://www.ibm.com");
+print '-- replace_regex';
+print replace_regex(strcat('Number is ', '1'), 'is (\d+)', 'was: \1');
```
Comment (kashwy) [inline tests/queries/0_stateless/02366_kql_func_string.sql:207]: \d  need a escape?

File: tests/queries/0_stateless/02366_kql_func_string.sql (hunk 1 above)
Comment (larryluogit) [reply tests/queries/0_stateless/02366_kql_func_string.sql:203]: are you sure? it works now without escape. And it works in Azure data explore.
adding escape makes no difference in this case.

File: tests/queries/0_stateless/02366_kql_func_ip.sql [hunk 3]
```diff
@@ -38,9 +38,87 @@ print '-- parse_ipv6(fe80::85d:e82c:9446:7994)';
...
+print format_ipv4_mask('192.168.1.1', 24);
+print format_ipv4_mask('192.168.1.1', 32);
+print format_ipv4_mask('192.168.1.1/24', -1) == '';
+print format_ipv4_mask('abc', 24) == '';
+print '-- parse_ipv6_mask()';
+print parse_ipv6_mask("127.0.0.1", 24) == '0000:0000:0000:0000:0000:ffff:7f00:0000';
+print parse_ipv6_mask("fe80::85d:e82c:9446:7994", 120) == 'fe80:0000:0000:0000:085d:e82c:9446:7900';
+-- print parse_ipv6_mask("192.168.255.255", 120) == '0000:0000:0000:0000:0000:ffff:c0a8:ff00';
+print parse_ipv6_mask("192.168.255.255/24", 124) == '0000:0000:0000:0000:0000:ffff:c0a8:ff00';
+print parse_ipv6_mask("255.255.255.255", 128) == '0000:0000:0000:0000:0000:ffff:ffff:ffff';
```
Comment (ltrk2) [inline tests/queries/0_stateless/02366_kql_func_ip.sql:93]: In these cases I believe that based on the specification only suffixes of 0-32 can be used in a meaningful way. This is reflected in the first example at https://docs.microsoft.com/en-us/azure/data-explorer/kusto/query/parse-ipv6-maskfunction.

File: tests/queries/0_stateless/02366_kql_func_string.sql (hunk 2 above)
Comment (larryluogit) [reply tests/queries/0_stateless/02366_kql_func_string.sql:207]: same as previous comments. it works without escape now.

File: tests/queries/0_stateless/02366_kql_func_ip.sql [hunk 4]
```diff
@@ -38,9 +38,87 @@ print '-- parse_ipv6(fe80::85d:e82c:9446:7994)';
...
+print parse_ipv6_mask("fe80::85d:e82c:9446:7994", 128) == 'fe80:0000:0000:0000:085d:e82c:9446:7994';
+print parse_ipv6_mask("fe80::85d:e82c:9446:7994/120", 124) == 'fe80:0000:0000:0000:085d:e82c:9446:7900';
+-- print parse_ipv6_mask("::192.168.255.255", 128) == '0000:0000:0000:0000:0000:ffff:c0a8:ffff';
+-- print parse_ipv6_mask("::192.168.255.255/24", 128) == '0000:0000:0000:0000:0000:ffff:c0a8:ff00';
```
Comment (ltrk2) [inline tests/queries/0_stateless/02366_kql_func_ip.sql:97]: In these cases I believe that based on the specification only suffixes of 0-32 can be used in a meaningful way. This is reflected in the first example at https://docs.microsoft.com/en-us/azure/data-explorer/kusto/query/parse-ipv6-maskfunction.

**Review by ltrk2 — CHANGES_REQUESTED** [review]

File: tests/queries/0_stateless/02366_kql_func_string.sql (hunk 1 above)
Comment (larryluogit) [reply tests/queries/0_stateless/02366_kql_func_string.sql:203]: I tested in Azure. The expression is correct. Adding escape return incorrect results in Azure data explore.

File: tests/queries/0_stateless/02366_kql_func_ip.sql (hunk 3 above)
Comment (ltrk2) [reply tests/queries/0_stateless/02366_kql_func_ip.sql:93]: I've just checked the Azure Data Explorer and these queries work there as you mentioned. This one, however: `print parse_ipv6_mask("127.0.0.1", 24)` does not produce the result stated at https://docs.microsoft.com/en-us/azure/data-explorer/kusto/query/parse-ipv6-maskfunction. In essence, I believe that their implementation is most likely non-compliant.

File: tests/queries/0_stateless/02366_kql_func_string.sql (hunk 1 above)
Comment (kashwy) [reply tests/queries/0_stateless/02366_kql_func_string.sql:203]: I think Azure data explore using Verbatim string literals, like **@"[^\w]+"**   so now escape needed.

`print trim_end(@"[^\w]+", strcat("- ","Te st", "1", "// $"));  `    and 
`print trim_end("[^\\w]+", strcat("- ","Te st", "1", "// $"))`

get same result:
```
"print_0": - Te st1,
```
`print trim_end("[^\w]+", strcat("- ","Te st", "1", "// $")); `      get result:
```

Error 

 I could not parse that, sorry.

 clientRequestId: KustoWebV2;3744a207-6a93-40df-a36a-ccadb65e817a
```

`print trim_end(@"[^\\w]+", strcat("- ","Te st", "1", "// $")); `   get empty

Comment (kashwy) [reply tests/queries/0_stateless/02366_kql_func_string.sql:203]: it all depends how a client tread the escape. in CH client, its good for both **"[^\\w]+"** and **"[^\w]+"**

so all good here

**Review by HeenaBansal2009 — APPROVED** [review]

**Review by kashwy — APPROVED** [review]

**Review by SadiHassan — APPROVED** [review]

File: tests/queries/0_stateless/02366_kql_func_ip.sql (hunk 3 above)
Comment (ltrk2) [reply tests/queries/0_stateless/02366_kql_func_ip.sql:93]: We discussed this with @larryluogit and came to the conclusion that the case I mentioned above should be disregarded and the function should work as if the mask was applied to the result of `parse_ipv6`.

**Pull request merged**